collect_app/src/test/java/org/odk/collect/android/http/OpenRosaServerClientFactoryTest.java

```diff
@@ -55,6 +66,23 @@ public void sendsOpenRosaHeaders() throws Exception {
         assertThat(request.getHeader("X-OpenRosa-Version"), equalTo("1.0"));
     }
 
+    @Test
+    public void sendsDateHeader() throws Exception {
```

Comment:
This test verifies three things -- that a date header is added, that the date header's format is correct and that the `getCurrentTime` method is called. I strongly agree with verifying the first one. I think the second is not a bad idea as documentation of intended behavior (especially that the date is always in English since that has caused problems). I'm not sure the third is all that meaningful. It's the third that requires introducing the `Clock` interface and that feels like dependency inversion gone a bit too far to me. Naturally, I can be convinced otherwise. 

Would you consider, instead of verifying that the date/time in the header is an exact match to a date/time produced by a mock, just checking the format of the header date/time? I think you could do that by parsing the date string with the desired formatter, formatting the resulting Date and ensuring it matches the original string.

Reply:
I can see your argument about it feeling like the test is checking a call to `getCurrentTime`. I didn't really intended it that way but more as a verification we are using the right time. Mockito maybe muddies that though. I'm not sure I like the idea of just checking the format. Looking at the [specs for a request](https://docs.opendatakit.org/openrosa-http/#http-requests) as far as I can see we need the `Date` header to be the "The date on the device" and also in that format. Given that it does feel like the things we need to drive out with a test are:

1. There is a date header
2. The date is in the right format
3. The date represents the correct point in time - the device's "now"

I can imagine an argument that we don't need the time to be dead on and that we could use fuzzy "time is within/near" matchers but I'm never quite comfortable introducing that non determinism when we can just treat time like the variant that it is and have it as a dependency on our object/function. 

I can totally see that the `Clock` seems like overkill for this one object but I'd be surprised if it wouldn't be useful in other objects and their tests (any of the time based Widgets for example). Do you think maybe using a "fake" would be nicer? This is a pattern I've actually used more than Mockito for something like this:

```java
class SettableClock implements Clock {

    private Date time;

    @Override
    public Date getTime() {
        if (time != null) {
            return time;
        } else {
            return new Date();
        }
    }

    public void setTime(Date time) {
        this.time = time;
    }
}

fakeClock.setTime(new Date());
subject = new OkHttpOpenRosaServerClientFactory(baseClient, clock);
```

Reply:
I guess there are two separate things that I don't love: that the mock doesn't make it really feel like the test really checks for current time and that the `Clock` dependency comes in the `OkHttpOpenRosaServerClientFactory` constructor.

For the first point, it's too bad we can't use `java.time`'s `Clock` (because of API level). It's of course very similar to what you have but the fact that the interface comes from the same package and that everything is designed to work together feels better. The fake vs. mock approach doesn't make a big difference to me on that dimension. Maybe it's the best that can be done.

For the second, I was hoping that perhaps the dependency could come in more locally but I guess that with `OkHttpOpenRosaServerClient` being private static, there's not really much that can be done there.

Ok, I give up, this is probably as good as can be.

Reply:
Yeah I kind of get you here. Having this "system level" abstraction here maybe feels a bit much. I think I'm maybe jumping ahead because I've used a `Clock` abstraction before but usually at more of a `Fragment`/`Activity` level. 

Maybe the way to do it at this level is just have `currentTime` come in as an argument to `makeRequest`. Seeing as we have this approved now I'll hold off but I'm happy to follow up with an attempt at something we can be happier with 😁

Reply:
> Having this "system level" abstraction here maybe feels a bit much.

Precisely.

Agree that an argument to `makeRequest` would be great. But you'd have to restructure things since you can't test `makeRequest` directly, right? Sounds good to keep thinking about these kinds of things for future improvements. It's good to have had the conversation so we get synced up on each other's taste.

I think this kind of passing around of low-level dependencies through a bunch of layers is how people end up feeling like they need a DI framework so I want to be mindful of it.

Reply:
It's pretty doable as `makeRequest` is called from the tests as it's the only public member on the `OpenRosaServerClient`! I'll play around with it when we get this in 👍